### PR TITLE
HW: part1 (Sergei_Ermushin@epam.com) 

### PR DESCRIPTION
**Pull request by Saver12**

### CODE REVIEW

File: src/main/java/spliterators/part1/exercise/RectangleSpliterator.java [hunk 1]
```diff
@@ -29,18 +28,51 @@ private RectangleSpliterator(int[][] array, int startOuterInclusive, int endOute
...
-        // TODO
-        throw new UnsupportedOperationException();
+        int length = endOuterExclusive - startOuterInclusive;
+        if (length <= 1) {
+            return null;
+        }
+
+        int middle = startOuterInclusive + length / 2;
+
+        final RectangleSpliterator newSpliterator = new RectangleSpliterator(array, startOuterInclusive, middle, 0);
```
Comment (senia-psm) [inline src/main/java/spliterators/part1/exercise/RectangleSpliterator.java]: Incorrect.

File: src/main/java/spliterators/part1/exercise/RectangleSpliterator.java [hunk 2]
```diff
@@ -29,18 +28,51 @@ private RectangleSpliterator(int[][] array, int startOuterInclusive, int endOute
...
-        // TODO
-        throw new UnsupportedOperationException();
+        if (startInnerInclusive < innerLength) {
+            action.accept(array[startOuterInclusive][startInnerInclusive]);
+            startInnerInclusive += 1;
+            return true;
+        } else if (startOuterInclusive < endOuterExclusive - 1) {
+            startOuterInclusive += 1;
+            startInnerInclusive = 0;
+            action.accept(array[startOuterInclusive][startInnerInclusive]);
```
Comment (senia-psm) [inline src/main/java/spliterators/part1/exercise/RectangleSpliterator.java]: Reduce code  duplication.

File: src/main/java/spliterators/part1/exercise/RectangleSpliterator.java [hunk 3]
```diff
@@ -29,18 +28,51 @@ private RectangleSpliterator(int[][] array, int startOuterInclusive, int endOute
...
+            return true;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void forEachRemaining(IntConsumer action) {
+        while (startOuterInclusive < endOuterExclusive) {
+            action.accept(array[startOuterInclusive][startInnerInclusive]);
```
Comment (senia-psm) [inline src/main/java/spliterators/part1/exercise/RectangleSpliterator.java]: Check `inner` boundary.

Add test for with an empty array.
